drivers/network/dd/b57xx/ndis.c

```diff
@@ -0,0 +1,574 @@
+    
+    if (Adapter->SendProducer[0].pPacketList != NULL)
+    {
+        NdisFreeMemory(Adapter->SendProducer[0].pPacketList,
+                       sizeof(PNDIS_PACKET) * Adapter->SendProducer[0].Count,
+                       0);
+    }
+    
+    if (Adapter->Status.pBlock != NULL)
+    {
+        NdisMFreeSharedMemory(Adapter->MiniportAdapterHandle,
+                              sizeof(B57XX_STATUS_BLOCK),
+                              FALSE,
+                              (PVOID)Adapter->Status.pBlock,
+                              Adapter->Status.HostAddress);
+    }
+    
+    if (Adapter->Statistics.pBlock != NULL)
+    {
+        NdisMFreeSharedMemory(Adapter->MiniportAdapterHandle,
+                              sizeof(B57XX_STATISTICS_BLOCK),
+                              FALSE,
+                              (PVOID)Adapter->Statistics.pBlock,
+                              Adapter->Statistics.HostAddress);
+    }
+    
+    /* Free MMIO */
+    if (Adapter->IoBase)
+    {
+        NdisMUnmapIoSpace(Adapter->MiniportAdapterHandle, Adapter->IoBase, Adapter->IoLength);
+    }
+
+    /* Destroy the adapter context */
+    NdisFreeMemory(Adapter, sizeof(B57XX_ADAPTER), 0);
+}
+
+NDIS_STATUS
+NTAPI
+MiniportSend(_In_ NDIS_HANDLE MiniportAdapterContext,
+             _In_ PNDIS_PACKET Packet,
+             _In_ UINT Flags)
+{
+    PB57XX_ADAPTER Adapter = (PB57XX_ADAPTER)MiniportAdapterContext;
+    NDIS_STATUS Status = NDIS_STATUS_PENDING;
+    PSCATTER_GATHER_LIST SGList = NDIS_PER_PACKET_INFO_FROM_PACKET(Packet,
+                                                                   ScatterGatherListPacketInfo);
+    LONG BusyDescriptors = Adapter->SendProducer[0].ProducerIndex -
+                           Adapter->SendProducer[0].ConsumerIndex;
+    
+    ASSERT(SGList != NULL);
+    ASSERT(SGList->NumberOfElements >= 1);
+    
+    NdisDprAcquireSpinLock(&Adapter->SendLock);
+    
+    if (BusyDescriptors < 0)
+    {
+        BusyDescriptors += Adapter->SendProducer[0].Count;
+    }
+    
+    if (Adapter->SendProducer[0].RingFull == TRUE ||
+        BusyDescriptors + SGList->NumberOfElements > Adapter->SendProducer[0].Count)
+    {
+        /* Force NDIS to resubmit the send */
+        Status = NDIS_STATUS_RESOURCES;
+        goto Exit;
+    }
+    
+    for (UINT i = 0; i < SGList->NumberOfElements; i++)
+    {
+        ASSERT(SGList->Elements[i].Length <= Adapter->MaxFrameSize);
+        
+        Adapter->SendProducer[0].pPacketList[Adapter->SendProducer[0].ProducerIndex] = Packet;
+
+        NICTransmitPacket(Adapter,
+                          SGList->Elements[i].Address,
+                          SGList->Elements[i].Length,
+                          i == SGList->NumberOfElements - 1 ? B57XX_SBD_PACKET_END : 0);
+    }
+    
+    if (Adapter->SendProducer[0].ProducerIndex == Adapter->SendProducer[0].ConsumerIndex)
+    {
+        NDIS_MinDbgPrint("Send ring is full\n");
+        Adapter->SendProducer[0].RingFull = TRUE;
```

Comment:
Is that flag still needed? It might be better to check the `BusyDescriptors` variable instead.